### PR TITLE
Syncer and Library abstractions

### PR DESCRIPTION
**Pull request by mcuadros**
- Library manages the local copy of the repositories.
- Syncer syncs a given revision of a repository to the local copy.

### CODE REVIEW

File: library.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	"gopkg.in/src-d/go-git.v4/storage"
+	"gopkg.in/src-d/go-git.v4/storage/filesystem"
+)
+
+var (
+	ErrRepositoryExists    = errors.NewKind("repository %s already exists")
+	ErrRepositoryNotExists = errors.NewKind("repository %s not exists")
+)
+
+// Library is controls the persistence of multiple git repositories.
```
Comment (smola) [inline library.go]: `s/is controls/controls/`

File: library.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+// Has return true if a repository with the given URL exists.
```
Comment (smola) [inline library.go]: `s/return/returns/`

File: library.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+func (l *Library) Has(url *RepositoryInfo) bool {
+	_, err := l.fs.Stat(l.repositoryPath(url))
```
Comment (smola) [inline library.go]: Do we really want to hide permission errors?

File: syncer.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
+package lookout
+
+import (
+	"context"
+	"fmt"
+
+	git "gopkg.in/src-d/go-git.v4"
```
Comment (smola) [inline syncer.go]: `s/git //`

File: syncer.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+// Sync syncs the local git repository to the given commit revision.
+func (s *Syncer) Sync(ctx context.Context, rev *CommitRevision) error {
+	r, err := s.l.GetOrInit(rev.Head.Repository)
+	if err != nil {
+		return err
+	}
+
+	opts := &git.FetchOptions{
```
Comment (smola) [inline syncer.go]: We don't need to execute a fetch if we already have the commit.

Comment (mcuadros) [reply syncer.go]: But we need to create the branch, so...

Comment (smola) [reply syncer.go]: Only if it does not exist already.

**Pull request merged**